games/mille/mille.6
.\"	$NetBSD: mille.6,v 1.13 2021/12/31 12:42:24 andvar Exp $
.\"
.\" Copyright (c) 1983, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)mille.6	8.3 (Berkeley) 6/1/94
.\"
.Dd June 1, 1994
.Dt MILLE 6
.Os
.Sh NAME
.Nm mille
.Nd play Mille Bornes
.Sh SYNOPSIS
.Nm mille
.Op Ar file
.Sh DESCRIPTION
.Nm Mille
plays a two-handed game reminiscent of
the Parker Brother's game of Mille Bornes with you.
The rules are described below.
If a file name is given on the command line,
the game saved in that file is started.
.Pp
When a game is started up,
the bottom of the score window will contain a list of commands.
They are:
.Bl -tag -width ".Ic P"
.It Ic P
Pick a card from the deck.
This card is placed in the
.Sq P
slot in your hand.
.It Ic D
Discard a card from your hand.
To indicate which card, type the number of the card in the hand
(or
.Sq P
for the just-picked card) followed
by a
.Aq RETURN
or
.Aq SPACE .
The
.Aq RETURN
or
.Aq SPACE
is required to allow recovery from typos
which can be very expensive, like discarding safeties.
.It Ic U
Use a card.
The card is again indicated by its number,
followed by a
.Aq RETURN
or
.Aq SPACE .
.It Ic O
Toggle ordering the hand.
By default off, if turned on it will sort the cards in your hand appropriately.
This is not recommended for the impatient on slow terminals.
.It Ic Q
Quit the game.
This will ask for confirmation, just to be sure.
Hitting
.Aq DELETE
(or
.Aq RUBOUT )
is equivalent.
.It Ic S
Save the game in a file.
If the game was started from a file,
you will be given an opportunity to save it on the same file.
If you don't wish to, or you did not start from a file,
you will be asked for the file name.
If you type a
.Aq RETURN
without a name, the save will be terminated and the game resumed.
.It Ic R
Redraw the screen from scratch.
The command ^L (control
.Sq L )
will also work.
.It Ic W
Toggle window type.
This switches the score window between the startup window
(with all the command names) and the end-of-game window.
Using the end-of-game window
saves time by eliminating the switch at the end of the game
to show the final score.
Recommended for hackers and other miscreants.
.El
.Pp
If you make a mistake, an error message will be printed
on the last line of the score window, and a bell will beep.
.Pp
At the end of each hand or game,
you will be asked if you wish to play another.
If not, it will ask you if you want to save the game.
If you do, and the save is unsuccessful,
play will be resumed as if you had said you wanted to play another hand/game.
This allows you to use the
.Dq Ic S
command to reattempt the save.
.Sh SEE ALSO
.Xr curses 3 ,
.Rs
.%T Screen Updating and Cursor Movement Optimization: A Library Package
.%A Ken Arnold
.Re
.Sh AUTHORS
.An Ken Arnold
.Pp
(The game itself is a product of Parker Brothers, Inc.)
.Sh CARDS
Here is some useful information.
The number in parentheses after the card name
is the number of that card in the deck:
.Bl -column "Speed Limit (3)" "End of Limit (6)" "Puncture Proof (1)"
.It Sy Hazard	Repair	Safety
.It Out of Gas (2)	Gasoline (6)	Extra Tank (1)
.It Flat Tire (2)	Spare Tire (6)	Puncture Proof (1)
.It Accident (2)	Repairs (6)	Driving Ace (1)
.It Stop (4)	Go (14)	Right of Way (1)
.It Speed Limit (3)	End of Limit (6)
.El
.Bd -centered
25 \- (10), 50 \- (10), 75 \- (10), 100 \- (12), 200 \- (4)
.Ed
.Sh RULES
.Ss Object :
The point of this game is to get a total of 5000 points in several hands.
Each hand is a race to put down exactly 700 miles before your opponent does.
Beyond the points gained by putting down milestones,
there are several other ways of making points.
.Ss Overview :
The game is played with a deck of 101 cards.
.Em Distance
cards represent a number of miles traveled.
They come in denominations of 25, 50, 75, 100, and 200.
When one is played,
it adds that many miles to the player's trip so far this hand.
.Em Hazard
cards are used to prevent your opponent from putting down Distance cards.
They can only be played if your opponent has a
.Em Go
card on top of the Battle pile.
The cards are
.Sy "Out of Gas" ,
.Sy "Accident" ,
.Sy "Flat Tire" ,
.Sy "Speed Limit" ,
and
.Sy "Stop" .
.Em Remedy
cards fix problems caused by Hazard cards played on you by your opponent.
The cards are
.Sy "Gasoline" ,
.Sy "Repairs" ,
.Sy "Spare Tire" ,
.Sy "End of Limit" ,
and
.Sy "Go" .
.Em Safety
cards prevent your opponent from putting specific Hazard cards on you
in the first place.
They are
.Sy "Extra Tank" ,
.Sy "Driving Ace" ,
.Sy "Puncture Proof" ,
and
.Sy "Right of Way" ,
and there are only one of each in the deck.
.Ss Board Layout :
The board is split into several areas.
From top to bottom, they are:
.Bl -ohang
.It Li SAFETY AREA
(unlabeled): This is where the safeties will be placed as they are played.
.It Li HAND :
These are the cards in your hand.
.It Li BATTLE :
This is the Battle pile.
All the Hazard and Remedy Cards are played here, except the
.Sy "Speed Limit"
and
.Sy "End of Limit"
cards.
Only the top card is displayed, as it is the only effective one.
.It Li SPEED :
The Speed pile.
The
.Sy "Speed Limit"
and
.Sy "End of Limit"
cards are played here
to control the speed at which the player is allowed to put down miles.
.It Li MILEAGE :
Miles are placed here.
The total of the numbers shown here is the distance traveled so far.
.El
.Ss Play :
The first pick alternates between the two players.
Each turn usually starts with a pick from the deck.
The player then plays a card, or if this is not possible or desirable,
discards one.
Normally, a play or discard of a single card constitutes a turn.
If the card played is a safety, however,
the same player takes another turn immediately.
.Pp
This repeats until one of the players reaches 700 points or the deck runs out.
If someone reaches 700, they have the option of going for an
.Em Extension ,
which means that the play continues until someone reaches 1000 miles.
.Ss Hazard and Remedy Cards :
Hazard Cards are played on your opponent's Battle and Speed piles.
Remedy Cards are used for undoing the effects of your opponent's nastiness.
.Bl -ohang
.It Sy Go No ( Green Light )
must be the top card on your Battle pile for you to play any mileage,
unless you have played the
.Sy "Right of Way"
card (see below).
.It Sy Stop
is played on your opponent's
.Sy Go
card to prevent them from playing mileage until they play a
.Sy Go
card.
.It Sy "Speed Limit"
is played on your opponent's Speed pile.
Until they play an
.Sy "End of Limit"
they can only play 25 or 50 mile cards, presuming their
.Sy Go
card allows them to do even that.
.It Sy "End of Limit"
is played on your Speed pile to nullify a
.Sy "Speed Limit"
played by your opponent.
.It Sy "Out of Gas"
is played on your opponent's
.Sy Go
card.
They must then play a
.Sy Gasoline
card, and then a
.Sy Go
card before they can play any more mileage.
.It Sy "Flat Tire"
is played on your opponent's
.Sy Go
card.
They must then play a
.Sy "Spare Tire"
card, and then a
.Sy Go
card before they can play any more mileage.
.It Sy "Accident"
is played on your opponent's
.Sy Go
card.
They must then play a
.Sy Repairs
card, and then a
.Sy Go
card before they can play any more mileage.
.El
.Ss Safety Cards :
Safety cards prevent your opponent
from playing the corresponding Hazard cards on you for the rest of the hand.
It cancels an attack in progress, and
.Em always entitles the player to an extra turn .
.Bl -ohang
.It Sy "Right of Way"
prevents your opponent from playing both
.Sy Stop
and
.Sy "Speed Limit"
cards on you.
It also acts as a permanent
.Sy Go
card for the rest of the hand, so you can play mileage
as long as there is not a Hazard card on top of your Battle pile.
In this case only, your opponent can play Hazard cards directly on a Remedy card
other than a Go card.
.It Sy "Extra Tank"
When played, your opponent cannot play an
.Sy "Out of Gas"
on your Battle Pile.
.It Puncture Proof
When played, your opponent cannot play a
.Sy "Flat Tire"
on your Battle Pile.
.It Sy "Driving Ace"
When played, your opponent cannot play an
.Sy Accident
on your Battle Pile.
.El
.Ss Distance Cards :
Distance cards are played when you have a
.Sy Go
card on your Battle pile,
or a Right of Way in your Safety area and are not stopped by a Hazard Card.
They can be played in any combination that totals exactly 700 miles,
except that
.Em you cannot play more than two 200 mile cards in one hand .
A hand ends whenever one player gets exactly 700 miles or the deck runs out.
In that case, play continues until neither someone reaches 700,
or neither player can use any cards in their hand.
If the trip is completed after the deck runs out, this is called
.Em "Delayed Action" .
.Ss Coup Fourr\('e :
This is a French fencing term for a counter-thrust move as part of a parry
to an opponent's attack.
In current French colloquial language it means a sneaky, underhanded blow.
In Mille Bornes, it is used as follows:
If an opponent plays a Hazard card,
and you have the corresponding Safety in your hand,
you play it immediately, even
.Em before
you draw.
This immediately removes the Hazard card from your Battle pile,
and protects you from that card for the rest of the game.
This gives you more points (see
.Sx Scoring
below).
.Ss Scoring :
Scores are totalled at the end of each hand,
whether or not anyone completed the trip.
The terms used in the Score window have the following meanings:
.Bl -ohang
.It Em Milestones Played :
Each player scores as many miles as they played before the trip ended.
.It Em Each Safety :
100 points for each safety in the Safety area.
.It Em All 4 Safeties :
300 points if all four safeties are played.
.It Em Each Coup Fourr\('e :
300 points for each Coup Fourr\('e accomplished.
.El
.Pp
The following bonus scores can apply only to the winning player.
.Bl -ohang
.It Em Trip Completed :
400 points bonus for completing the trip to 700 or 1000.
.It Em Safe Trip :
300 points bonus for completing the trip without using any 200 mile cards.
.It Em Delayed Action :
300 points bonus for finishing after the deck was exhausted.
.It Em Extension :
200 points bonus for completing a 1000 mile trip.
.It Em Shut-Out :
500 points bonus for completing the trip
before your opponent played any mileage cards.
.El
.Pp
Running totals are also kept for the current score for each player
for the hand
.Em ( "Hand Total" ) ,
the game
.Em ( "Overall Total" ) ,
and number of games won
.Em ( Games ) .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2855 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 9.0 to NetBSD 10.0:
	openldap: Import 2.4.48. [christos 20190808]
	boot(8): GPT and RAIDframe support for x86 bootstrap [manu 20190818]
	dhcpcd(8): Import dhcpcd-8.0.3 [roy 20190821]
	inet6: Fix sending RTM_MISS for unresolvable addresses and restore
		ND6_LLINFO_WAITDELETE so that any send at this point can
		return EHOSTDOWN to match inet behaviour. [roy 20190901]
	inet: Fix sending RTM_MISS for unresolvable addresses. [roy 20190901]
	dhcpcd(8): Import dhcpcd-8.0.4 [roy 20190904]
	bind: Import version 9.14.5. [christos 20190905]
	resolvconf(8): Import openresolv-3.9.2 [roy 20190908]
	boot(8): multiboot 2 support, Xen can now boot from EFI [manu 20190913]
	dhcpcd(8): Import dhcpcd-8.0.6 [roy 20190913]
	kernel: Add vHCI, a driver that allows to send and receive USB
		packets from userland. [maxv 20190914]
	mos(4): Ported driver for MosChip MCS7730/7830/7832 USB ethernet.
		[mrg 20190920]
	gcc(1): Import GCC 8.3. [mrg 20191001]
	libpcap: Import 1.9.1. [christos 20191001]
	tcpdump(8): Import 4.9.3. [christos 20191001]
	rescue(8): Moved into a separate, rescue-only set. [maya 20191001]
	byacc: Update to 20190617. [christos 20191006]
	tzdata updates from 2019a to 2019c (incl 2019b) [kre 20191008]
	services(5): Update to version 2019-10-04 from IANA.
		[sevan 20191009]
	dhcpcd(8): Import dhcpcd-8.1.0 [roy 20191011]
	OpenSSH: Import 8.1. [christos 20191012]
	acpi(4): Updated ACPICA to 20190816. [christos 20191015]
	sup: removed, use ftp://ftp.astron.com/pub/sup [christos 20191015]
	dhcpcd(8): Import dhcpcd-8.1.1 [roy 20191016]
	evbarm: Add support for the NanoPi R1 and Duo2. [bad 20191016]
	bind: Import version 9.14.7. [christos 20191017]
	libc: Changed the iconv(3) function signature to the POSIX variation
		[kamil 20191024]
	nct(4): Add Nuvoton NCT5104D GPIO driver. [ad 20191025]
	vio9p(4): Add virtio-9p front-end driver. [ozaki-r 20191028]
	mount_9p(8): Enable to mount a filesystem exported via virtio-9p.
		[ozaki-r 20191028]
	evbarm: Update TI AM335x SoC support to use FDT based configuration.
		[jmcneill 20191028]
	evbarm: Update TI OMAP3 SoC support to use FDT based configuration.
		[jmcneill 20191101]
	amd64: Add support for kCSan - Kernel Concurrency Sanitizer.
		[maxv 20191105]
	tmux(1): Imported 2.9a. [christos 20191112]
	dhcpcd(8): Import 8.1.2. [roy 20191113]
	amd64: Add support for kMSan - Kernel Memory Sanitizer.
		[maxv 20191114]
	qat(4): Add opencrypto driver for Intel QuickAssist.
		Atom C2XXX, C3XXX, Xeon D-21XX and D-15XX, C62X chipsets and
		QuickAssist Adapter 8960/8970.
		[hikaru 20191120]
	select(2), poll(2): Performance and scalability improvements
		[ad 20191121]
	amd64, i386: Performance and scalability improvements for TLB
		shootdown [ad 20191121]
	bind: Import version 9.14.8. [christos 20191127]
	terminfo: Import 20190609 [christos 20191127]
	onewire(4), owtemp(4): Improve reliability and reduce CPU overhead
		slightly. [ad 20191130]
	scheduler: Improve performance and interactivity. [ad 20191201]
	vfs: Reduce false sharing and lock overhead during normal operation.
		[ad 20191201]
	evbarm: Added driver for Allwinner Crypto Engine random number
		generator [riastradh 20191209]
	ixl(4): Ported driver for Intel Ethernet 700 series
		[yamaguchi 20191210]
	acpi(4): Updated ACPICA to 20191213. [christos 20191214]
	uvm: Replace red-black tree page lookup with radix tree lookup,
		including tagging and gang lookup of pages.  Original work
		done by yamt@. [ad 20191214]
	nsd: Import 4.2.4. [christos 20191215]
	unbound(8): Import 1.9.6. [christos 20191215]
	openpam(3): update to 20190224 (tabebuia) [christos 20191215]
 	kerberos(8): Update to Heimdal-7.7.0 [christos 20191215]
	file(1): Upgraded to 5.38. [christos 20191216]
	uvm: Replace global counters with per-CPU counters, significantly
		reducing system time on workloads that are VM-system heavy.
		[ad 20191216]
	anxdp(4): Add driver for Analogix DisplayPort core found in
		Rockchip RK3399. [jakllsch 20191219]
	net80211: Require SSID configuration to associate with an open AP.
		[jakllsch 20191219]
	dhcpcd(8): Import dhcpcd-8.1.4 [roy 20191220]
	uvm: Rewrite the page allocator to be more efficient and somewhat CPU
		topology aware.  Add preliminary NUMA support.  [ad 20191227]
	crunchgen(1): Added the ability to pass variables in the submake
		command line, and removed all command like flags that are
		not needed anymore [christos 20191229]
	uvm: Reduce lock contention by making the maintentance of page
		replacement state more concurrent. [ad 20191231]
	aq(4): Add Aquantia 10G network adapter driver [ryo 20200101]
	dhcpcd(8): Import dhcpcd-8.1.5 [roy 20200103]
	tmux(1): Imported 3.0a. [christos 20200106]
	amd64, i386: Replace the global PV hash table with radix trees,
		yielding significant performance and scalability benefits.
		[ad 20200112]
	scheduler: Make the scheduler topology aware.  It understands and
		makes good use of HyperThreading/SMT, spreads the load evenly

games/mille/mille.c
/*	$NetBSD: mille.c,v 1.20 2011/08/31 16:24:56 plunky Exp $	*/

/*
 * Copyright (c) 1982, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1982, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)mille.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD: mille.c,v 1.20 2011/08/31 16:24:56 plunky Exp $");
#endif
#endif /* not lint */

#include <time.h>
#include "mille.h"
#include <signal.h>

/*
 * @(#)mille.c	1.3 (Berkeley) 5/10/83
 */

int
main(int ac, char *av[])
{
	bool	restore;

	/* Revoke setgid privileges */
	setgid(getgid());

	if (strcmp(av[0], "a.out") == 0) {
		outf = fopen("q", "w");
		setbuf(outf, NULL);
		Debug = TRUE;
	}
	restore = FALSE;
	switch (ac) {
	  case 2:
		rest_f(av[1]);
		restore = TRUE;
	  case 1:
		break;
	  default:
		printf("usage: milles [ restore_file ]\n");
		exit(1);
		/* NOTREACHED */
	}
	Play = PLAYER;
	if (!initscr())
		errx(0, "couldn't initialize screen");
	delwin(stdscr);
	stdscr = Board = newwin(BOARD_Y, BOARD_X, 0, 0);
	Score = newwin(SCORE_Y, SCORE_X, 0, 40);
	Miles = newwin(MILES_Y, MILES_X, 17, 0);
#ifdef attron
	idlok(Board, TRUE);
	idlok(Score, TRUE);
	idlok(Miles, TRUE);
#endif
	leaveok(Score, TRUE);
	leaveok(Miles, TRUE);
	clearok(curscr, TRUE);
#ifndef PROF
	srandom((long) time(NULL));
#else
	srandom(0);
#endif
	cbreak();
	noecho();
	signal(SIGINT, rub);
	for (;;) {
		if (!restore || (Player[PLAYER].total >= 5000
		    || Player[COMP].total >= 5000)) {
			if (Player[COMP].total < Player[PLAYER].total)
				Player[PLAYER].games++;
			else if (Player[COMP].total > Player[PLAYER].total)
				Player[COMP].games++;
			Player[COMP].total = 0;
			Player[PLAYER].total = 0;
		}
		do {
			if (!restore)
				Handstart = Play = other(Handstart);
			if (!restore || On_exit) {
				shuffle();
				init();
			}
			newboard();
			if (restore)
				mvwaddstr(Score, ERR_Y, ERR_X, Initstr);
			prboard();
			do {
				domove();
				if (Finished)
					newscore();
				prboard();
			} while (!Finished);
			check_more();
			restore = On_exit = FALSE;
		} while (Player[COMP].total < 5000
		    && Player[PLAYER].total < 5000);
	}
}

/*
 *	Routine to trap rubouts, and make sure they really want to
 * quit.
 */
void
rub(int dummy __unused)
{
	(void)signal(SIGINT, SIG_IGN);
	if (getyn(REALLYPROMPT))
		die(0);
	(void)signal(SIGINT, rub);
}

/*
 *	Time to go beddy-by
 */
void
die(int code)
{

	(void)signal(SIGINT, SIG_IGN);
	if (outf)
		fflush(outf);
	mvcur(0, COLS - 1, LINES - 1, 0);
	endwin();
	exit(code);
}
